storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    pruner::state_store::{generics::StaleNodeIndexSchemaTrait, StateMerklePruner},
+    schema::{
+        db_metadata::{DbMetadataSchema, DbMetadataValue},
+        jellyfish_merkle_node::JellyfishMerkleNodeSchema,
+    },
+    utils::get_progress,
+};
+use anyhow::Result;
+use aptos_jellyfish_merkle::StaleNodeIndex;
+use aptos_schemadb::{schema::KeyCodec, SchemaBatch, DB};
+use aptos_types::transaction::{AtomicVersion, Version};
+use std::{
+    cmp::max,
+    marker::PhantomData,
+    sync::{atomic::Ordering, Arc},
+};
+
+pub(in crate::pruner) struct StateMerkleMetadataPruner<S> {
+    metadata_db: Arc<DB>,
+    next_version: AtomicVersion,
+    _phantom: PhantomData<S>,
+}
+
+impl<S: StaleNodeIndexSchemaTrait> StateMerkleMetadataPruner<S>
+where
+    StaleNodeIndex: KeyCodec<S>,
+{
+    pub(in crate::pruner) fn new(metadata_db: Arc<DB>) -> Self {
+        Self {
+            metadata_db,
+            next_version: AtomicVersion::new(0),
```

Comment:
Man I have to say this field is useless.. looking at the code it's always been overwritten by `current_progress`.

Reply:
Not really, for most of the time it's always be larger than current_progress, and it's overwritten by either the second smallest version that is larger than current_progress (the smallest version >= current_progress get pruned), or the target_version (if we've already reach the end of stale index in this round)

Reply:
okay.. needed a vc session to understand what you are saying, but seems legit 😞